### DISCUSSION TITLE
difference between 2.1.14 and 3.4.2

### DISCUSSION THREAD

**Discussion by abhishek-ksolve** [moj-analytical-services/splink#891] [Q&A]
When running in spark, what's the difference between splink 2.1.14 vs 3.4.2? also, I need to run 3.4.2 on local any running source code.

**Comment by RobinL** [comment]
There have been significant improvements to the API and improvements in speed in version 3 vs version 2, see the following links:

https://www.robinlinacre.com/splink_3/
https://github.com/moj-analytical-services/splink/releases/tag/v3.3.0. (and subsequent release notes).

Splink (version 2 or 3) should work in Spark 2.4 or above but version 3 or above is recommended.

**Reply by abhishek-ksolve** [reply]
hy RobinL I run SparkLinker locally it run successfully, But when I submit the same job on AWS glue it gives an error while running this function "linker.estimate_u_using_random_sampling(target_rows=5e5)". pls find the attachment.
Also, I use an additional jar name "scala-udf-similarity-0.0.10.jar" for spark.udf.registerJavaFunction("jaro_winkler", "uk.gov.moj.dash.linkage.JaroWinklerSimilarity", types.DoubleType())
Here my two questions are 1.) what's the cause of the error, pls find the attachment.
2) Is there any standard way to use this jar on aws glue, because locally it runs without a jar?
[aws_error.txt](https://github.com/moj-analytical-services/splink/files/10004732/aws_error.txt)

**Reply by RobinL** [reply]
Unable to infer schema often means you're reading a file with zero rows.  This error could occur if one of your input files has zero rows. another possibility is a Splink intermediate output has zero rows.  It might help to increase target rows to (say) 10 million.

**Reply by abhishek-ksolve** [reply]
Sorry Robin got same error.
Basically, I am using a single dataset/data frame containing a row of 10 to 12 k, And I have tried to run the target_rows on 
all 1e7,1e8, and 1e9, but every time on AWS glue it shows the same error of "AnalysisException: Unable to infer schema for Parquet. It must be specified manually."
Do you have any examples or references for submitting the SparkLinker job on AWS glue? Because with the same dataset I am able to get output from the local setup.
pls, find the attachment below there is the syntax that I am currently following
[code.txt](https://github.com/moj-analytical-services/splink/files/10006221/code.txt)
.

**Reply by RobinL** [reply]
The script looks ok to me.

You'll want to change:

`    "blocking_rules_to_generate_predictions": ["l.Country = r.Country"and"l.Description = r.Description]"`
to
`    "blocking_rules_to_generate_predictions": ["l.Country = r.Country", "l.Description = r.Description]"`

but i don't think that's the source of the error.

I think perhaps you need to set the checkpoint directory to an s3 location i.e. change
`spark.sparkContext.setCheckpointDir("./tmp_checkpoints")`
to 
`spark.sparkContext.setCheckpointDir("s3://xxxx/tmp_checkpoints")`

**Reply by abhishek-ksolve** [reply]
Thank you Robin for your help. Yes it's resolved by passing the tmp_checkpoints location to s3

**Reply by abhishek-ksolve** [reply]
Required keyword: 'expression' missing for <class 'sqlglot.expressions.Mul'>. Line 8, Col: 50.
  ctors
    ), 
__splink__df_predict as (
    select
    log2(cast(0.0026117515763011434 as double) * [4m)[0m as match_weight,
    ((cast(0.0026117515763011434 as double) * )/(1+(cast(0.0026117515763011434 as 

Required keyword: 'expression' missing for <class 'sqlglot.expressions.Mul'>. Line 9, Col: 47.
  0.0026117515763011434 as double) * ) as match_weight,
    ((cast(0.0026117515763011434 as double) * [4m)[0m/(1+(cast(0.0026117515763011434 as double) * ))) as match_probability,
    `unique_id_l`,`unique_id_

Required keyword: 'expression' missing for <class 'sqlglot.expressions.Mul'>. Line 9, Col: 93.
  weight,
    ((cast(0.0026117515763011434 as double) * )/(1+(cast(0.0026117515763011434 as double) * [4m)[0m)) as match_probability,
    `unique_id_l`,`unique_id_r`,`store_name_l`,`store_name_r`,`address_l`,`

**Comment by abhishek-ksolve** [comment]
@RobinL Need to know: I directly run this
linker = SparkLinker(df, settings)
results = linker.predict(threshold_match_probability=0.8) 
and got the same output as earlier, then what's the main use of these functions:

linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.6)
linker.estimate_u_using_random_sampling(target_rows=1e7)
training_session_fname_sname = linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule)

**Reply by RobinL** [reply]
This is covered in the [tutorial](https://moj-analytical-services.github.io/splink/demos/00_Tutorial_Introduction.html)
